### PR TITLE
Fixes #32991 - PageLayout to use PF4 components

### PR DESCRIPTION
**Pull request by MariaAga**
Starting to create an index page that would be used as a template for tables.
A screenshot that shows all the available buttons(create, export, documentation), models page doesn't have them all so its fake and the pr doesn't add them :)
Blocked by https://github.com/theforeman/foreman/pull/8990
Screenshot from 2021-12-22 15-31-37

### CODE REVIEW

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/ActionButtons.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/ActionButtons.js:28]: it would be awesome to allow here to provide a custom element as a prop or in a composable approach

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/ActionButtons.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/ActionButtons.js:39]: same here, consumers might prefer to pass the `dropdownItems` to the component,
so we should give it as an option

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js:35]: though it will suit almost all index pages, we might want to call it `PageLayout` as it can fit also to other pages imho

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]:
```suggestion
  const memoOptions = useMemo(() => ({ ...apiOptions, params }), [params]);
```

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]: do we really need to prefix them with `custom` ? :D

File: webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js]: I don't think we need to use a separate PF4 dir.
eventually, all components will be PF4, and I would vote for removing the old `PageLayout`

File: webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js]: Is it OK to set this PR to be blocked on #8990 which removed some code from the models page?

**Review by Ron-Lavi — COMMENTED** [review]
Thanks @MariaAga !
I am really happy to see this PR, we are moving in the right direction :D
I would suggest waiting for #8990 which also add some changes to the models page
and to add the pagination to the toolbar.

the code looks great 🥳  left just a few comments though I would suggest replacing the PageLayout component instead of creating a new dir under PF4 dir as soon enough there won't be any PF3 components around 🤞🏼

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/ActionButtons.js
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/ActionButtons.js:39]: I think it's better that if consumers want something custom I will add a `toolbarComponent` in the `TableIndexPage` component and not overcomplicate the `ActionButtons`

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js:35]: I think we will add a different page layout to non table pages, as they don't need search/action buttons/pagination

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]: I didn't want there to be a confusion that the consumer **has** to provide the link for the search/button/actions

File: webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js]: If were keeping the `TableIndexPage` name I can move it out :)

File: webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js]: np

**Comment by MariaAga** [comment]
I want to add the pagination in a different pr to keep them small :)

**Comment by Ron-Lavi** [comment]
The pagination PR got in, can you please rebase?

**Comment by Ron-Lavi** [comment]
I would like to try and use it in the new reports index page and in the puppet tab, currently, the design is:
https://marvelapp.com/prototype/2b4h8bg6/screen/84598222
Screenshot (39)

I think we should implement just the toolbar in this PR, e.g:

Screenshot (41)

And the page header could be implemented differently on each page, e.g:
Screenshot (42)

In this way, we could also reuse the toolbar on the new host page where the title isn't needed for example.

**Comment by MariaAga** [comment]
Rebased and added `cutsomToolbarItems` so a component can add it's own "special" buttons like in reports

**Comment by MariaAga** [comment]
@Ron-Lavi Can you take another look?:)

**Comment by MariaAga** [comment]
Fixed ruby test (hopefully)

**Comment by Ron-Lavi** [comment]
Awesome work @MariaAga !!
Would it be OK if we wait after branching?
I could also update https://github.com/theforeman/foreman_host_reports/pull/21 based on your work

**Comment by MariaAga** [comment]
Sounds great, thanks!

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]: why do we need to trigger the autocomplete#getResults here? shouldn't it be handled in the autocomplete component?

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]: why do we need to trigger `setAutocomplete` on bookmark click?

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]: In case someone wants to trigger a modal, can they override this behavior? e.g `onClick`

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js:115]: should it be key based so we could override the action items? e.g:
```js
{
  create: { ... },
  export: { ... },
  help: { ... },
  ...customActionButtons,
}
```

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]: The section here seems to have too much space between the elements and search bar is a bit shifted to the right:
Screenshot (58)

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]: can the consumer pass `onSearch` callback so in the parent component they could manage the search change as well and trigger an API call in the table for example.

File: webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js:5]: if I am not mistaken, this layout is not only for tables, should it leave also under /common
everything is changing to PF4 so moving into this folder and later we would probably switch to common won't make sense, how about:

```suggestion
import IndexPageLayout from '../../common/PageLayout/IndexPageLayout';
```

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]: Gonna address it here:
https://projects.theforeman.org/issues/34546

File: webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js:5]: I think as we add search, create new and export button here this page should be only used for tables

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js:115]: Devs can just pass `creatable={false}` and `exportable`,`hasHelpPage` are false by default so they wont be shown

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]: This page does the api call using:
```
  const {
    response: { search: apiSearchQuery, can_create: canCreate },
    status = STATUS.PENDING,
  } = useAPI('get', apiUrl, memoOptions);
```
So if a dev want a custom onSearch they should probably not use this page, but create a custom page for their specifications

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]: ^ was resolved so `setAutocomplete` is deleted

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]: Changed to onClick and then devs can add navigate or open action if needed

**Review by Ron-Lavi — COMMENTED** [review]
Overall looks great @MariaAga !
Can you also add a storybook to show how to use it?
Maybe @jeremylenz @Andrewgdewar will have more opinions about this? Thanks

**Comment by jeremylenz** [comment]
I think our storybook build is broken right now due to our inability to upgrade Webpack.. :cry:

File: webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/PF4/TableIndexPage/TableIndexPage.js]: can you try again after #8713 was merged?

File: webpack/assets/javascripts/react_app/routes/Models/ModelsPage/components/ModelsPageContent.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/routes/Models/ModelsPage/components/ModelsPageContent.js:25]: looks like the search bar update the URL on search,
but the table content doesn't get updated

Comment (MariaAga) [reply webpack/assets/javascripts/react_app/routes/Models/ModelsPage/components/ModelsPageContent.js:25]: After the useApi hook update I needed to use `useEffect` for `setAPIOptions` to update the table, should work now

**Review by Ron-Lavi — APPROVED** [review]
Thanks @MariaAga !
The page works well and looks better 👍🏼 
There is still some work to do with the new layout wrapper,
and also to change the table to PF4 one, though I'm OK with doing it in steps,
we should also document how to use it properly and hopefully the storybook will be fixed soon 👼🏼

**Pull request merged**
